libvmi/driver/kvm/kvm_events.c

```diff
@@ -0,0 +1,835 @@
+    //          CS
+    libvmi_event->x86_regs->cs_base = kvmi_event->event.common.arch.sregs.cs.base;
+    libvmi_event->x86_regs->cs_limit = kvmi_event->event.common.arch.sregs.cs.limit;
+    libvmi_event->x86_regs->cs_sel = kvmi_event->event.common.arch.sregs.cs.selector;
+    //          DS
+    libvmi_event->x86_regs->ds_base = kvmi_event->event.common.arch.sregs.ds.base;
+    libvmi_event->x86_regs->ds_limit = kvmi_event->event.common.arch.sregs.ds.limit;
+    libvmi_event->x86_regs->ds_sel = kvmi_event->event.common.arch.sregs.ds.selector;
+    //          SS
+    libvmi_event->x86_regs->ss_base = kvmi_event->event.common.arch.sregs.ss.base;
+    libvmi_event->x86_regs->ss_limit = kvmi_event->event.common.arch.sregs.ss.limit;
+    libvmi_event->x86_regs->ss_sel = kvmi_event->event.common.arch.sregs.ss.selector;
+    //          ES
+    libvmi_event->x86_regs->es_base = kvmi_event->event.common.arch.sregs.es.base;
+    libvmi_event->x86_regs->es_limit = kvmi_event->event.common.arch.sregs.es.limit;
+    libvmi_event->x86_regs->es_sel = kvmi_event->event.common.arch.sregs.es.selector;
+    //          FS
+    libvmi_event->x86_regs->fs_base = kvmi_event->event.common.arch.sregs.fs.base;
+    libvmi_event->x86_regs->fs_limit = kvmi_event->event.common.arch.sregs.fs.limit;
+    libvmi_event->x86_regs->fs_sel = kvmi_event->event.common.arch.sregs.fs.selector;
+    //          GS
+    libvmi_event->x86_regs->gs_base = kvmi_event->event.common.arch.sregs.gs.base;
+    libvmi_event->x86_regs->gs_limit = kvmi_event->event.common.arch.sregs.gs.limit;
+    libvmi_event->x86_regs->gs_sel = kvmi_event->event.common.arch.sregs.gs.selector;
+    //      VCPU
+    libvmi_event->vcpu_id = kvmi_event->event.common.vcpu;
+}
+
+static status_t
+process_cb_response(
+    vmi_instance_t vmi,
+    event_response_t response,
+    vmi_event_t *libvmi_event,
+    struct kvmi_dom_event *kvmi_event,
+    void *rpl,
+    size_t rpl_size)
+{
+    kvm_instance_t *kvm = kvm_get_instance(vmi);
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!kvm || !kvm->kvmi_dom) {
+        errprint("%s: invalid kvm or kvmi handles\n", __func__);
+        return VMI_FAILURE;
+    }
+    if (!libvmi_event || !kvmi_event || !rpl) {
+        errprint("%s: invalid libvmi/kvmi/rpl handles\n", __func__);
+        return VMI_FAILURE;
+    }
+#endif
+    // loop over all possible responses
+    for (uint32_t i = VMI_EVENT_RESPONSE_NONE+1; i <=__VMI_EVENT_RESPONSE_MAX; i++) {
+        event_response_t candidate = 1u << i;
+        if (response & candidate) {
+            switch (candidate) {
+                default:
+                    errprint("%s: KVM - unhandled event reponse %u\n", __func__, candidate);
+                    break;
+            }
+        }
+    }
+    // the rpl struct should be like this
+    //    struct {
+    //        struct kvmi_vcpu_hdr hdr;
+    //        struct kvmi_event_reply common;
+    //        // Event specific structs
+    //        ...
+    //    };
```

Comment:
This type of structure being passed around is really odd. Instead of it being casted and passed around, just allocate each header individually and pass the required pointers to the functions that needs them.

Reply:
The issue I was attempting to solve is that the reply might contain event specific fields.
And we need to set these event specified in the event handler.

Example with `process_register()`

~~~C
    // reply struct
     struct {
         struct kvmi_vcpu_hdr hdr;
         struct kvmi_event_reply common;
         struct kvmi_event_cr_reply cr;
     } rpl = {0};


     // TODO: how can the callback specifiy a new value for the MSR ?
     // libvmi_event.reg_event.xxx
     rpl.cr.new_val = kvmi_event->event.cr.new_value;

     // and then calling the generic "process_callback_response" function, passing rpl pointer.
~~~

I was following the example provided in `kvmi`:
https://github.com/KVM-VMI/kvm/blob/kvmi/tools/kvm/kvmi/kvmi-test.c#L83
cc @mdontu, @adlazar 

> nstead of it being casted and passed around, just allocate each header individually and pass the required pointers to the functions that needs them.

So, if I understand correctly,  I would do something like this
~~~C
// allocate header
struct kvmi_vcpu_hdr hdr;
// allocate common
 struct kvmi_event_reply common;
// allocate event specific struct if need
struct kvmi_event_cr_reply cr;

int size = sizeof(hdr) + sizeof(common) + sizeof(cr);

// pass them to reply function
process_callback_response(..., &hdr, &common, (void*)&cr, size );
~~~

:thinking:

Reply:
Well, it looks to me `kvmi_vcpu_hdr` and `kvmi_event_reply` are standard headers all events would use, so I would update the prototype of process_callback_response to take the pointers to those structs directly. If the `kvmi_event_cr_reply` is variable based on the event type, it would be fine to pass that as a void* and then cast it when needed to its respective type in the part of process_callback_response that needs it. Does that make sense?

Reply:
I think we can even remove this part
~~~C
   struct kvmi_vcpu_hdr *hdr = (struct kvmi_vcpu_hdr *)rpl;
     hdr->vcpu = kvmi_event->event.common.vcpu;

     struct kvmi_event_reply *common = (struct kvmi_event_reply *)( hdr + 1 );
     common->event = kvmi_event->event.common.event;
     common->action = KVMI_EVENT_ACTION_CONTINUE;
~~~

and make the `process_xxxx` functions do the reply.

Example:
~~~C
process_register(vmi_instance_t vmi, struct kvmi_dom_event *kvmi_event)
{
  // handle register event ...
  // call callback
  event_response_t response = call_event_callback(vmi, libvmi_event);
  // prepare reply struct
    struct {
         struct kvmi_vcpu_hdr hdr;
         struct kvmi_event_reply common;
         struct kvmi_event_cr_reply cr;
     } rpl = {0};

   int action = process_cb_response()

   rpl.hdr.vcpu = kvmi_event->event.common.vcpu;
   rpl.common.event = kvmi_event->event.common.event;
   rpl.action = action;

   // do the reply
   if (kvm->libkvmi.kvmi_reply_event(kvm->kvmi_dom, kvmi_event->seq, rpl, sizeof(rpl)))
         return VMI_FAILURE;

   return VMI_SUCCESS;
}

static int process_cb_response()
{
    // loop over all possible responses
     for (uint32_t i = VMI_EVENT_RESPONSE_NONE+1; i <=__VMI_EVENT_RESPONSE_MAX; i++) {
         event_response_t candidate = 1u << i;
         if (response & candidate) {
             switch (candidate) {
                 default:
                     errprint("%s: KVM - unhandled event reponse %u\n", __func__, candidate);
                     break;
             }
         }
     }

     return KVMI_EVENT_ACTION_CONTINUE; // or KVMI_EVENT_ACTION_DENY, ...
}
~~~

What do you think

Reply:
So progressing on the implementation, I would like to add support for
event responses like
- `VMI_EVENT_RESPONSE_SET_EMUL_READ_DATA`
- `VMI_EVENT_RESPONSE_SET_EMUL_INSN`

However, there is a catch.
Only the pagefault events can offer emulation in replies:
~~~C
struct kvmi_event_pf_reply {
	__u64 ctx_addr;
	__u32 ctx_size;
	__u8 singlestep;
	__u8 rep_complete;
	__u16 padding;
	__u8 ctx_data[256];
};
~~~

So there are 2 solutions to implement this

# Solution 1

Handle `VMI_EVENT_RESPONSE_SET_EMUL_READ_DATA` and `VMI_EVENT_RESPONSE_SET_EMUL_INSN` in the same common function: `process_cb_response`.

However, since we are passing a `void *rpl`, we have to cast and do pointer arithmetic, similar to
~~~C
    struct kvmi_vcpu_hdr *hdr = (struct kvmi_vcpu_hdr *)rpl;
    //  

    struct kvmi_event_reply *common = (struct kvmi_event_reply *)( hdr + 1 );
    // 


   case VMI_EVENT_RESPONSE_SET_EMUL_READ_DATA:
          struct kvmi_event_pf_reply *pf_reply = (struct kvmi_event_pf_reply *)( common + 1 );
          // handle emulation response
          memcpy(pf_reply.ctx_data, ....)
~~~

What I don't like with this solution is that we are making ugly casts, and aren't doing any sanity check.
If a `interrupt` callback sets this event response, the `* rpl` struct will be overwritten here.

Here comes solution 2

# Solution 2

What I would prefer is declaring a second function, with the only goal to handle this emulation event response, and which will be called from the `process_pagefault` handler
~~~C
            // call user callback
            event_response_t response = call_event_callback(vmi, libvmi_event);

            // handle emulation reply requests
            if (VMI_FAILURE == process_cb_response_emulate(vmi, response, libvmi_event, &rpl))
                return VMI_FAILURE;
~~~


@tklengyel which one do you prefer ?
Thanks.